File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        new_head = ListNode()
        itr = new_head
        self.merge(itr, list1, list2)
        
        return new_head.next
    
    def merge(self, itr, list1, list2):
        if list1 and list2:
            if list1.val <= list2.val:
                itr.next = list1
                self.merge(itr.next, list1.next, list2)
            else:
                itr.next = list2
                self.merge(itr.next, list1, list2.next)
        elif list1:
            itr.next = list1
            self.merge(itr.next, list1.next, list2)
        elif list2:
            itr.next = list2
            self.merge(itr.next, list1, list2.next)
            
        

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode idx1 = list1;
        ListNode idx2 = list2;
        ListNode head = new ListNode(0);
        ListNode cur = head;
        
        if (list1 == null) 
            return list2;
        else if (list2 == null)
            return list1;
        
        while (idx1 != null && idx2 != null) {
            if (idx1.val < idx2.val) {
                cur.next = idx1;
                idx1 = idx1.next;
            } else {
                cur.next = idx2;
                idx2 = idx2.next;
            }
            cur = cur.next;
        }
        
        while (idx1 != null) {
            cur.next = idx1;
            idx1 = idx1.next;
            cur = cur.next;
        }
        while (idx2 != null) {
            cur.next = idx2;
            idx2 = idx2.next;
            cur = cur.next;
        }        
        
        return head.next;

    }
}